### PR TITLE
logalloc, reader_concurrency_semaphore: cooperate on OOM kills

### PR DESCRIPTION
**Pull request by denesb**
Consider the following code snippet:
```c++
future<> foo() {
    semaphore.consume(1024);
}

future<> bar() {
    return _allocating_section([&] {
        foo();
    });
}
```

If the consumed memory triggers the OOM kill limit, the semaphore will throw `std::bad_alloc`. The allocating section will catch this, bump std reserves and retry the lambda. Bumping the reserves will not do anything to prevent the next call to `consume()` from triggering the kill limit. So this cycle will repeat until std reserves are so large that ensuring the reserve fails. At this point LSA gives up and re-throws the `std::bad_alloc`. Beyond the useless time spent on code that is doomed to fail, this also results in expensive LSA compaction and eviction of the cache (while trying to ensure reserves).
Prevent this situation by throwing a distinct exception type which is derived from `std::bad_alloc`. Allocating section will not retry on seeing this exception.
A test reproducing the bug is also added.

Fixes: #15278

### CODE REVIEW

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3985/

Dtest fix here: https://github.com/scylladb/scylla-dtest/pull/3607.

**Comment by avikivity** [comment]
> `logalloc::allocating_section`, retries the code run through it, with increasing reserves, until it either succeeds or the reserve cannot be allocated. This interacts badly with the reader concurrency semaphore's OOM kill feature, which also uses an `std::bad_alloc` to terminate the current read. If this happens inside an allocating section, as it is the case in the cache reader, this results in LSA trying to increase reserves until it fails. Increasing reserves doesn't help to improve the semaphore's memory pressure, so this results in a tail-spin until reserves are so large they cannot be allocated. Beyond the useless time spend on a read that is doomed to fail already, this also results in expensive compaction and eviction of the cache. Prevent this situation by throwing a distinct exception type which is derived from `std::bad_alloc`. Allocating section will not retry on seeing this exception. A test reproducing the bug is also added.
> 
> Fixes: #15278

I don't understand the problem yet. Won't the bad_alloc be caught be LSA, which can then work to satisfy the read? It won't leak to reader_concurrency_semaphore until we really do run out of memory.

**Comment by michoecho** [comment]
> I don't understand the problem yet. Won't the bad_alloc be caught be LSA, which can then work to satisfy the read? It won't leak to reader_concurrency_semaphore until we really do run out of memory.

Rephrasing what Botond already said in the cover letter: the bug is about a situation where the `allocating_sections` is *outside* the `reader_concurrency_semaphore` call.
`reader_concurrency_semaphore` has its own memory limit, and throws `bad_alloc` when this limit is exceeded. This is intended to kill the current read.
However, if this happens inside `allocating_section`, the section catches the exception and attempts to retry the operation instead of letting the read die as intended. And this retry can't succeed, because the semaphore limits the memory held by readers — it is not affected by LSA reclaim at all.
In other words, the problem is that the same exception type is used for two different memory-limiting mechanisms, and they accidentally catch each other's exceptions.

**Comment by denesb** [comment]
Cover-letter updated. Looks like English is too ambiguous a language. I instead used a language we all understand best: C++.

**Comment by denesb** [comment]
Re-kicked CI, the dtest changes have landed.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4005/

https://github.com/scylladb/scylladb/issues/15602
https://github.com/scylladb/scylladb/issues/15530

**Comment by denesb** [comment]
CI failures seem unrelated on first look but they are suspiciously persistent. I will need to investigate.

**Comment by denesb** [comment]
> CI failures seem unrelated on first look but they are suspiciously persistent. I will need to investigate.

They were unrelated after all, we had some problems with the build farms.

**Comment by denesb** [comment]
@avikivity ping.
